File: examples/LearningHadoop/MapReduceRecSysJava/src/main/java/com/example/UserVectorToCooccurrenceMapper.java
package com.example;

import org.apache.hadoop.mapreduce.Mapper;
import org.apache.mahout.math.VarIntWritable;
import org.apache.mahout.math.VarLongWritable;
import org.apache.mahout.math.Vector;
import org.apache.mahout.math.VectorWritable;

import java.io.IOException;
import java.util.Iterator;

/**
 * Created by andryw on 10/02/15.
 */
public class UserVectorToCooccurrenceMapper extends Mapper<VarLongWritable,VectorWritable,VarIntWritable,VarIntWritable> {

    public void map(VarLongWritable userID, VectorWritable userVector, Context context) throws IOException, InterruptedException {
        Iterator<Vector.Element> it = userVector.get().iterateNonZero();
        while (it.hasNext()){
            int index1 = it.next().index();
            Iterator<Vector.Element> it2 = userVector.get().iterateNonZero();
            while (it2.hasNext()){
                int index2 = it2.next().index();
                context.write(new VarIntWritable(index1), new VarIntWritable(index2));
            }

        }
    }
}

File: nMiners/src/nMiners/API/UserVectorToCooccurrenceMapper.scala
package API
import Utils._

import org.apache.hadoop.io.{Text, LongWritable}
import org.apache.hadoop.mapred.{Reporter, OutputCollector, Mapper, MapReduceBase}
import org.apache.mahout.math.Vector.Element
import org.apache.mahout.math.{VarIntWritable, VectorWritable, VarLongWritable}

/**
 * Created by arthur on 07/04/15.
 */
class UserVectorToCooccurrenceMapper extends MapReduceBase with Mapper[VarLongWritable,VectorWritable,VarIntWritable,VarIntWritable]{

  def map( userID : VarLongWritable, userVector:VectorWritable , output: OutputCollector[VarIntWritable, VarIntWritable], reporter: Reporter) = {

    val it  = userVector get() nonZeroes() iterator()
    it.foreach((item: Element) => {
      val it2  = userVector get() nonZeroes() iterator()
      val index1 = item.index()
      it2.foreach((item2:Element)=> {
        output.collect(index1,item2.index())
      })
    })
  }
  }
